static/app/utils/useProjects.tsx

```diff
@@ -1,13 +1,315 @@
+  }
+
+  return {results: data, hasMore, nextCursor};
+}
 
 /**
  * Provides projects from the ProjectStore
+ *
+ * This hook also provides a way to select specific project slugs, and search
+ * (type-ahead) for more projects that may not be in the project store.
+ *
+ * NOTE: Currently ALL projects are always loaded, but this hook is designed
+ * for future-compat in a world where we do _not_ load all projects.
  */
-function useProjects() {
-  const {projects, loading} = useLegacyStore(ProjectsStore);
+function useProjects({limit, slugs, orgId: propOrgId}: Options = {}) {
+  const api = useApi();
+
+  const {organization} = useLegacyStore(OrganizationStore);
+  const store = useLegacyStore(ProjectsStore);
+
+  const orgId = propOrgId ?? organization?.slug;
+
+  const storeSlugs = new Set(store.projects.map(t => t.slug));
+  const slugsToLoad = slugs?.filter(slug => !storeSlugs.has(slug)) ?? [];
+  const shouldLoadSlugs = slugsToLoad.length > 0;
+
+  const [state, setState] = useState<State>({
+    initiallyLoaded: !store.loading && !shouldLoadSlugs,
+    fetching: shouldLoadSlugs,
+    hasMore: null,
+    lastSearch: null,
+    nextCursor: null,
+    fetchError: null,
+  });
+
+  const slugsRef = useRef<Set<string> | null>(null);
+
+  // Only initialize slugsRef.current once and modify it when we receive new
+  // slugs determined through set equality
+  if (slugs !== undefined) {
+    if (slugsRef.current === null) {
+      slugsRef.current = new Set(slugs);
+    }
+
+    if (
+      slugs.length !== slugsRef.current.size ||
+      slugs.some(slug => !slugsRef.current?.has(slug))
+    ) {
+      slugsRef.current = new Set(slugs);
+    }
+  }
+
+  async function loadProjectsBySlug() {
+    if (orgId === undefined) {
+      // eslint-disable-next-line no-console
+      console.error('Cannot use useProjects({slugs}) without an organization in context');
+      return;
+    }
+
+    setState({...state, fetching: true});
+    try {
+      const {results, hasMore, nextCursor} = await fetchProjects(api, orgId, {
+        slugs: slugsToLoad,
+        limit,
+      });
+
+      const fetchedProjects = uniqBy([...store.projects, ...results], ({slug}) => slug);
+      ProjectActions.loadProjects(fetchedProjects);
+
+      setState({
+        ...state,
+        hasMore,
+        fetching: false,
+        initiallyLoaded: !store.loading,
+        nextCursor,
+      });
+    } catch (err) {
+      console.error(err); // eslint-disable-line no-console
+
+      setState({
+        ...state,
+        fetching: false,
+        initiallyLoaded: !store.loading,
+        fetchError: err,
+      });
+    }
+  }
+
+  async function handleSearch(search: string) {
+    const {lastSearch} = state;
+    const cursor = state.nextCursor;
+
+    if (search === '') {
+      return;
+    }
+
+    if (orgId === undefined) {
+      // eslint-disable-next-line no-console
+      console.error('Cannot use useProjects.onSearch without an organization in context');
```

Comment:
I know we do this in the other hooks, but instead of console error should we just send these events to sentry?

Reply:
I believe console.error bubbles up to sentry.

But I should verify